feat(framework): add script to generate tests

Diogo21Costa opened the pull request:
## PR Description

This PR introduces a Python script that automates the generation of tests for the test framework. The need for this script arose because the project can contain multiple C code files with different tests, and users may want to run only a subset of these tests. Hence, this script has been developed to create a set of macros that enable users to selectively choose which tests to run.

To use the script, follow the code below:

```bash
python3 codegen.py -d tests -o bao-tests/src/testf_entry.c
```

When running the script with the specified arguments, it will search for test files in the tests directory and generate macros to handle test selection. The output will be written to bao-tests/src/testf_entry.c.

Please review this PR and provide your feedback.

## Type of change

- **feat**: introduces a new functionality
  - Logical unit: test generation automation script

## Checklist:

- [x] The changes generate no new warnings when building the project. If so, I have justified above.
- [x] I have run the CI checkers before submitting the PR to avoid unnecessary runs of the workflow.

ESCristiano commented on codegen.py:
```diff
@@ -0,0 +1,127 @@
+This script is used to generate Bao Project tests code.
+It searches for C source files with 'BAO_TEST' markers and creates corresponding test functions.
+"""
+
```
@Diogo21Costa you should standardize the comments, either use **# Comment X** or **""" Comment X """**.  And please limit the column width in the whole file to 80 characters.

Diogo21Costa replied:
Thank you for your review! I wanted to clarify the difference between the comment styles used in the code:

1. The triple-quoted (""" Comment X """) comment at the start is a module docstring. It provides an overview of the module's purpose.
2. The single-line comments (# Comment X) explain specific lines of code or steps for developers. In this case, they serve to specify the license.

Both styles serve distinct purposes, enhancing readability. Maintaining this differentiation helps convey both the module's intention and code specifics effectively.

In addition to addressing the comment, I've adjusted the code's line width to ensure it falls within the 80-character limit. Thank you for your consideration!

ESCristiano replied:
Thank you @Diogo21Costa for the clarification. It makes sense now! No more concerns from my side. I'll approve.

miguelafsilva5 commented:
The PR should follow the template format. I couldn't find a better link, but you can find the close PR with the template here [https://github.com/bao-project/bao-ci/pull/28#issue-1418662001](url)

miguelafsilva5 commented:
Wouldn't it be better if, instead of writing a completly new file from hard coded strings, the script opened a template/skeleton .c file and copied it and added the new function calls? This would allow for easier modification to the base file.

Diogo21Costa commented:
> Wouldn't it be better if, instead of writing a completly new file from hard coded strings, the script opened a template/skeleton .c file and copied it and added the new function calls? This would allow for easier modification to the base file.

Your suggestion makes a lot of sense. Instead of creating an entirely new file using hardcoded strings, I've now modified the script to open a template/skeleton .c file, copy it, and then add the new function calls. This will make it much easier to make changes to the base file.

Thanks for the feedback! The changes have been implemented and can be viewed in commit [6deee6a](https://github.com/bao-project/bao-tests/pull/8/commits/6deee6a1862e51d63e4c589cc348092b875fab06).

Diogo21Costa commented:
> The PR should follow the template format. I couldn't find a better link, but you can find the close PR with the template here [bao-project/bao-ci#28](https://github.com/bao-project/bao-ci/pull/28)

@miguelafsilva5 Thank you for bringing this up. I've now updated the PR description to align with the one you provided. I appreciate your feedback!
